### PR TITLE
Int8 review

### PR DESCRIPTION
**Pull request by ZhennanQin**
## Description ##
(Brief description on what this PR is about)

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [MXNET-$JIRA_ID], where $JIRA_ID refers to the relevant [JIRA issue](https://issues.apache.org/jira/projects/MXNET/issues) created (except PRs with tiny changes)
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [ ] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [ ] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [ ] Feature1, tests, (and when applicable, API doc)
- [ ] Feature2, tests, (and when applicable, API doc)

## Comments ##
- If this change is a backward incompatible change, why must this change be made.
- Interesting edge cases to note here

### CODE REVIEW

File: example/quantization/imagenet_gen_qsym_mkldnn.py [hunk 1]
```diff
@@ -0,0 +1,218 @@
...
+        logger.info('Saving params into file at %s' % fname)
+    save_dict = {('arg:%s' % k): v.as_in_context(cpu()) for k, v in arg_params.items()}
+    save_dict.update({('aux:%s' % k): v.as_in_context(cpu()) for k, v in aux_params.items()})
+    mx.nd.save(fname, save_dict)
+
+
+if __name__ == '__main__':
+    parser = argparse.ArgumentParser(description='Generate a calibrated quantized model from a FP32 model with mkldnn support')
+    parser.add_argument('--model', type=str, choices=['imagenet1k-resnet-152', 'imagenet1k-inception-bn'],
+                        help='currently only supports imagenet1k-resnet-152 or imagenet1k-inception-bn')
```
Comment (pengzhao-intel) [inline example/quantization/imagenet_gen_qsym_mkldnn.py]: We need to support more models rather than only the original two models

File: example/quantization/imagenet_gen_qsym_mkldnn.py [hunk 2]
```diff
@@ -0,0 +1,218 @@
...
+def save_params(fname, arg_params, aux_params, logger=None):
+    if logger is not None:
+        logger.info('Saving params into file at %s' % fname)
+    save_dict = {('arg:%s' % k): v.as_in_context(cpu()) for k, v in arg_params.items()}
+    save_dict.update({('aux:%s' % k): v.as_in_context(cpu()) for k, v in aux_params.items()})
+    mx.nd.save(fname, save_dict)
+
+
+if __name__ == '__main__':
+    parser = argparse.ArgumentParser(description='Generate a calibrated quantized model from a FP32 model with mkldnn support')
```
Comment (pengzhao-intel) [inline example/quantization/imagenet_gen_qsym_mkldnn.py]: mkldnn to MKL-DNN

File: example/quantization/imagenet_gen_qsym_mkldnn.py [hunk 3]
```diff
@@ -0,0 +1,218 @@
...
+                        help='path of the calibration dataset')
+    parser.add_argument('--image-shape', type=str, default='3,224,224')
+    parser.add_argument('--data-nthreads', type=int, default=60,
+                        help='number of threads for data decoding')
+    parser.add_argument('--num-calib-batches', type=int, default=10,
+                        help='number of batches for calibration')
+    parser.add_argument('--exclude-first-conv', action='store_true', default=True,
+                        help='excluding quantizing the first conv layer since the'
+                             ' number of channels is usually not a multiple of 4 in that layer'
+                             ' which does not satisfy the requirement of cuDNN')
```
Comment (pengzhao-intel) [inline example/quantization/imagenet_gen_qsym_mkldnn.py]: update the help comments. This is a new file and specific for MKL-DNN. 
Don't related with 'cuDNN'

File: example/quantization/imagenet_gen_qsym_mkldnn.py [hunk 4]
```diff
@@ -0,0 +1,218 @@
...
+                             ' 3. entropy: calculate KL divergence of the fp32 output and quantized output for optimal'
+                             ' thresholds. This mode is expected to produce the best inference accuracy of all three'
+                             ' kinds of quantized models if the calibration dataset is representative enough of the'
+                             ' inference dataset.')
+    parser.add_argument('--quantized-dtype', type=str, default='uint8',
+                        choices=['int8', 'uint8'],
+                        help='quantization destination data type for input data')
+    parser.add_argument('--disable-requantize', type=bool, default=True,
+                        help='If disable requantize, the OP needed requantize'
+                             ' will output int8 directly and hence requantize '
```
Comment (pengzhao-intel) [inline example/quantization/imagenet_gen_qsym_mkldnn.py]: output is `int8` or `uint8`?

File: example/quantization/imagenet_gen_qsym_mkldnn.py [hunk 5]
```diff
@@ -0,0 +1,218 @@
...
+    logger = logging.getLogger('logger')
+    logger.setLevel(logging.INFO)
+
+    logger.info('shuffle_dataset=%s' % args.shuffle_dataset)
+
+    calib_mode = args.calib_mode
+    logger.info('calibration mode set to %s' % calib_mode)
+
+    # download calibration dataset
+    if calib_mode != 'none':
```
Comment (pengzhao-intel) [inline example/quantization/imagenet_gen_qsym_mkldnn.py:114]: how to use local dataset and parameter file?

File: example/quantization/imagenet_gen_qsym_mkldnn.py [hunk 6]
```diff
@@ -0,0 +1,218 @@
...
+    check_call(_LIB.MXGenBackendSubgraph(c_str(backend), sym.handle, ctypes.byref(out)))
+    sym = Symbol(out)
+
+    # get batch size
+    batch_size = args.batch_size
+    logger.info('batch size = %d for calibration' % batch_size)
+
+    # get number of batches for calibration
+    num_calib_batches = args.num_calib_batches
+    if calib_mode != 'none':
```
Comment (pengzhao-intel) [inline example/quantization/imagenet_gen_qsym_mkldnn.py]: Add the message for `none` mode too

File: example/quantization/imagenet_gen_qsym_mkldnn.py [hunk 7]
```diff
@@ -0,0 +1,218 @@
...
+
+    # get number of threads for decoding the dataset
+    data_nthreads = args.data_nthreads
+
+    # get image shape
+    image_shape = args.image_shape
+
+    exclude_first_conv = args.exclude_first_conv
+    excluded_sym_names = []
+    if args.model == 'imagenet1k-resnet-152':
```
Comment (pengzhao-intel) [inline example/quantization/imagenet_gen_qsym_mkldnn.py:145]: support more models, at least resnet-50 v1, SSD-vgg16, maskRCNN.

**Review by pengzhao-intel — COMMENTED** [review]
some comments for usability

File: include/mxnet/c_api.h [hunk 8]
```diff
@@ -1502,22 +1502,23 @@ MXNET_DLL int MXSymbolInferType(SymbolHandle sym,
...
+* \param quantized_dtype the quantized destination type for input data.
+* \param disable_requantize whether disable requantize OP during quantization
+* \param calib_quantize whether calibrate quantize op with offline calibration data.
+*/
+int MXQuantizeSymbol(SymbolHandle sym_handle, SymbolHandle *ret_sym_handle,
+                     const mx_uint num_excluded_symbols,
+                     const char **excluded_symbols,
+                     const mx_uint num_offline, const char **offline_params,
+                     const char *quantized_dtype, const bool disable_requantize,
+                     const bool calib_quantize);
```
Comment (pengzhao-intel) [inline include/mxnet/c_api.h]: How about wrapping all parameters by a structure so that the API interface doesn't need to change in the future?

File: python/mxnet/contrib/quantization.py [hunk 9]
```diff
@@ -69,34 +70,46 @@ def _quantize_params(qsym, params):
...
+            output = name[: - len('_min')] + "_output"
+            if output in th_dict:
+                quantized_params[name] = ndarray.array([th_dict[output][1]])
     return quantized_params
 
-
 def _quantize_symbol(sym, excluded_symbols=None, offline_params=None,
-                     quantized_dtype='int8'):
+                     quantized_dtype='int8', disable_requantize=False,
+                     calib_quantize_op=False):
```
Comment (pengzhao-intel) [inline python/mxnet/contrib/quantization.py]: As previous, how to make the interface easy to extend in the future?

**Review by pengzhao-intel — COMMENTED** [review]
is it possible to make API interface back compatible?

File: example/quantization/imagenet_gen_qsym_mkldnn.py (hunk 1 above)
Comment (ZhennanQin) [reply example/quantization/imagenet_gen_qsym_mkldnn.py]: Yes, this can be done when resnet50 and sdd-vgg pre-train model ready.

File: example/quantization/imagenet_gen_qsym_mkldnn.py (hunk 5 above)
Comment (ZhennanQin) [reply example/quantization/imagenet_gen_qsym_mkldnn.py:114]: Currently not support. We can add this feature.

File: example/quantization/imagenet_gen_qsym_mkldnn.py (hunk 7 above)
Comment (ZhennanQin) [reply example/quantization/imagenet_gen_qsym_mkldnn.py:145]: Sure.

File: include/mxnet/c_api.h (hunk 8 above)
Comment (ZhennanQin) [reply include/mxnet/c_api.h]: It's hard to use structure here as it's interface between python and C. Python don't know c structure alignment, it's hard to pack structure from python.

File: Makefile [hunk 10]
```diff
@@ -66,8 +66,8 @@ $(warning "USE_MKL2017 is deprecated. We will switch to USE_MKLDNN.")
 endif
 
 ifeq ($(USE_MKLDNN), 1)
-	MKLDNNROOT = $(ROOTDIR)/3rdparty/mkldnn/install
-	MKLROOT = $(ROOTDIR)/3rdparty/mkldnn/install
+	MKLDNNROOT = $(ROOTDIR)/3rdparty/mkldnn/build/install
+	MKLROOT = $(ROOTDIR)/3rdparty/mkldnn/build/install
```
Comment (TaoLv) [inline Makefile:70]: Why change this? If it's not related to this fusion/quantization, I suggest to move it to another PR for change.

File: example/quantization/imagenet_gen_qsym_mkldnn.py [hunk 11]
```diff
@@ -0,0 +1,219 @@
+# Licensed to the Apache Software Foundation (ASF) under one
```
Comment (TaoLv) [inline example/quantization/imagenet_gen_qsym_mkldnn.py:1]: Please elaborate why we need a mkldnn specific quantization script. (Maybe need add doc string to this file.)

File: example/quantization/imagenet_gen_qsym_mkldnn.py [hunk 12]
```diff
@@ -0,0 +1,219 @@
...
+    if calib_mode != 'none':
+        download_calib_dataset('http://data.mxnet.io/data/val_256_q90.rec', args.calib_dataset)
+
+        # download model
+        prefix, epoch = download_model(model_name=args.model, logger=logger)
+        sym, arg_params, aux_params = mx.model.load_checkpoint(prefix, epoch)
+
+    out = SymbolHandle()
+    backend = "MKLDNN"
+    check_call(_LIB.MXGenBackendSubgraph(c_str(backend), sym.handle, ctypes.byref(out)))
```
Comment (TaoLv) [inline example/quantization/imagenet_gen_qsym_mkldnn.py]: Is this an open API which can be called by the user?

File: include/mxnet/c_api.h [hunk 13]
```diff
@@ -1527,13 +1528,18 @@ MXNET_DLL int MXQuantizeSymbol(SymbolHandle sym_handle,
...
 MXNET_DLL int MXSetCalibTableToQuantizedSymbol(SymbolHandle qsym_handle,
                                                const mx_uint num_layers,
                                                const char** layer_names,
                                                const float* low_quantiles,
                                                const float* high_quantiles,
-                                               SymbolHandle* ret_sym_handle);
+                                               SymbolHandle* ret_sym_handle,
+                                               const bool disable_requantize);
+
+MXNET_DLL int MXGenBackendSubgraph(const char *backend, SymbolHandle sym_handle,
```
Comment (TaoLv) [inline include/mxnet/c_api.h]: Put SymbolHandle as the first parameter? Same style with other APIs.

File: include/mxnet/ndarray.h [hunk 14]
```diff
@@ -702,6 +702,11 @@ class NDArray {
    * It's used by FullyConnected right now.
    */
   NDArray MKLDNNDataReshape(const TShape &shape) const;
+
+   /*!
+   * \ Fix mkldnn memory descriptor mismatch from NDArray.
+   */
+  void UpdateMKLDNNMemDesc();
```
Comment (TaoLv) [inline include/mxnet/ndarray.h:709]: If this is not related to fusion/quantization, suggest to create another PR to fix it.

File: include/mxnet/op_attr_types.h [hunk 15]
```diff
@@ -300,6 +300,14 @@ using FQuantizedOp = std::function<nnvm::NodePtr (const NodeAttrs& attrs)>;
  */
 using FNeedRequantize = std::function<bool (const NodeAttrs& attrs)>;
 
+/*!
+ * \brief Register a function to determine if the input of a quantized operator
+ * needs to be quantized. This is usually used for the quantized operators
+ * which can handle fp32 inputs directly.
+ */
```
Comment (TaoLv) [inline include/mxnet/op_attr_types.h:307]: Could you elaborate more about when an operator has several inputs? How to decide which input will be skipped?

File: mkldnn.mk [hunk 16]
```diff
@@ -47,7 +47,7 @@ $(MKLDNN_LIBFILE):
 
 mkldnn_clean:
 	$(RM) -r 3rdparty/mkldnn/build
-	$(RM) -r 3rdparty/mkldnn/install/*
+	$(RM) -r $(MKLDNNROOT)
```
Comment (TaoLv) [inline mkldnn.mk:50]: Create another PR for this change?

File: python/mxnet/contrib/quantization.py [hunk 17]
```diff
@@ -69,34 +70,46 @@ def _quantize_params(qsym, params):
             quantized_params[name+'_max'] = vmax
         elif name in params:
             quantized_params[name] = params[name]
+        elif name.endswith(('_min')):
+            output = name[: - len('_min')] + "_output"
+            if output in th_dict:
+                quantized_params[name] = ndarray.array([th_dict[output][0]])
+        elif name.endswith(('_max')):
+            output = name[: - len('_min')] + "_output"
```
Comment (TaoLv) [inline python/mxnet/contrib/quantization.py:78]: nit: _max?

File: python/mxnet/contrib/quantization.py [hunk 18]
```diff
@@ -69,34 +70,46 @@ def _quantize_params(qsym, params):
...
     """Given a symbol object representing a neural network of data type FP32,
     quantize it into a INT8 network.
 
     Parameters
     ----------
     sym : Symbol
         FP32 neural network symbol.
-    excluded_symbols : list of symbols
-        Nodes in the network that users do not want to replace with a symbol of INT8 data type.
+    excluded_sym_names : list of strings
```
Comment (TaoLv) [inline python/mxnet/contrib/quantization.py:92]: Not match to the parameter name in the paramter list.

File: src/executor/attach_op_execs_pass.cc [hunk 19]
```diff
@@ -159,9 +159,13 @@ class StatefulComputeExExecutor : public OpExecutor {
     op_ctx.run_ctx = rctx;
 #if MXNET_USE_MKLDNN == 1
     InvalidateOutputs(out_array, req);
-    CreateDefaultInputs(in_array, &in_array_fallback);
-    fcompute_(state_, op_ctx, in_array_fallback, req, out_array);
-    return;
+    // TODO(alex): (MXNET-847) Remove this fallback feature after subgraph implemented
```
Comment (TaoLv) [inline src/executor/attach_op_execs_pass.cc:162]: Another PR?

File: src/ndarray/ndarray.cc [hunk 20]
```diff
@@ -449,24 +449,6 @@ void NDArray::Chunk::SetMKLMem(const TShape &shape, int dtype) {
   mkl_mem_.reset(new MKLDNNMemory(pd, shandle.dptr));
 }
 
-/*
```
Comment (TaoLv) [inline src/ndarray/ndarray.cc:452]: Why remove this?

File: src/operator/subgraph/mkldnn/mkldnn_conv.cc [hunk 21]
```diff
@@ -0,0 +1,670 @@
...
+  DType *bias_ptr = no_bias ? nullptr : bias->data().dptr<DType>();
+  DType *gamma_ptr = gamma.Reorder2Default().data().dptr<DType>();
+  DType *beta_ptr = beta.Reorder2Default().data().dptr<DType>();
+  DType *mean_ptr = mean.Reorder2Default().data().dptr<DType>();
+  DType *var_ptr = variance.Reorder2Default().data().dptr<DType>();
+  DType *update_weight_ptr = update_weight.data().dptr<DType>();
+  DType *update_bias_ptr = update_bias.data().dptr<DType>();
+  size_t channel = gamma.shape()[0];
+  size_t offset = weight->shape()[1] * weight->shape()[2] * weight->shape()[3];
+#pragma omp parallel for
```
Comment (TaoLv) [inline src/operator/subgraph/mkldnn/mkldnn_conv.cc]: Get available thread number from mxnet::engine::OpenMP.

File: src/operator/subgraph/mkldnn/mkldnn_conv.cc [hunk 22]
```diff
@@ -0,0 +1,670 @@
...
+  }
+}
+
+static void ConvFusionFallBackCompute() {
+  LOG(FATAL) << "Don't know how to do ConvFusionFallBackCompute!";
+}
+
+static void ConvolutionFusionComputeExCPU(const MKLDNNConvFullParam &full_param,
+                                          const OpContext &ctx,
+                                          MKLDNNConvForward *fwd,
```
Comment (TaoLv) [inline src/operator/subgraph/mkldnn/mkldnn_conv.cc:163]: const?

File: src/operator/subgraph/mkldnn/mkldnn_conv.cc [hunk 23]
```diff
@@ -0,0 +1,670 @@
...
+    os << ")";
+    throw dmlc::ParamError(os.str());
+  }
+  auto subgraph_sym = attrs->subgraphs[0];
+  DFSVisit(subgraph_sym->outputs, [&](const nnvm::NodePtr &node) {
+    if (node->is_variable()) return;
+    auto &node_name = node->op()->name;
+    if (node_name == "BatchNorm") {
+      CHECK_EQ(param_.full_conv_param.mkldnn_param.with_bn, true);
+      CHECK(param_.bn_param.get() == nullptr);
```
Comment (TaoLv) [inline src/operator/subgraph/mkldnn/mkldnn_conv.cc:422]: CHECK_EQ?

File: src/operator/subgraph/mkldnn/mkldnn_conv_property.cc [hunk 24]
```diff
@@ -0,0 +1,241 @@
...
+};
+
+class SgMKLDNNConvProperty : public SubgraphProperty {
+ public:
+  SgMKLDNNConvProperty() {
+    disable_all = dmlc::GetEnv("MXNET_DISABLE_MKLDNN_FUSION", 0);
+    disable_conv_bn = dmlc::GetEnv("MXNET_DISABLE_MKLDNN_FUSION_CONV_BN", 0);
+    disable_conv_relu = dmlc::GetEnv("MXNET_DISABLE_MKLDNN_FUSION_CONV_RELU", 0);
+    disable_conv_sum = dmlc::GetEnv("MXNET_DISABLE_MKLDNN_FUSION_CONV_SUM", 0);
+
```
Comment (TaoLv) [inline src/operator/subgraph/mkldnn/mkldnn_conv_property.cc]: disable_all = disable_conv_bn && disable_conv_relu && disable_conv_sum ?

File: Makefile (hunk 10 above)
Comment (ZhennanQin) [reply Makefile:70]: Yes, it could be separated into another PR, I put it here just because it's a minor change.

File: example/quantization/imagenet_gen_qsym_mkldnn.py (hunk 11 above)
Comment (ZhennanQin) [reply example/quantization/imagenet_gen_qsym_mkldnn.py:1]: Firstly, this is the agreement from meeting with Wujun and Zhengda. Secondly, each backend will do quantization differently, and we need to provide a easy-to-use script to help customer try mkldnn solution quickly. Thirdly, we will add resnet50v1, sdd-vgg, and maskrcnn in this scripts in future, and we can't guarantee those models can works well on other backends. So it's better to have our own scritps.

File: example/quantization/imagenet_gen_qsym_mkldnn.py (hunk 12 above)
Comment (ZhennanQin) [reply example/quantization/imagenet_gen_qsym_mkldnn.py]: Yes, I designed so. User may want to get a fused model in python level, and do more experiment based on that, eg. save to json, view graph change, evaluate quantization work.

File: include/mxnet/c_api.h (hunk 13 above)
Comment (ZhennanQin) [reply include/mxnet/c_api.h]: Make sense.

File: include/mxnet/ndarray.h (hunk 14 above)
Comment (ZhennanQin) [reply include/mxnet/ndarray.h:709]: It's related to quantization. We need this function to update mkldnn mem descriptor from a quantized conv+sum.

File: include/mxnet/op_attr_types.h (hunk 15 above)
Comment (ZhennanQin) [reply include/mxnet/op_attr_types.h:307]: OK. I will add more comments to help understand.

File: python/mxnet/contrib/quantization.py (hunk 17 above)
Comment (ZhennanQin) [reply python/mxnet/contrib/quantization.py:78]: Good catch, although the functionality is the same.

File: python/mxnet/contrib/quantization.py (hunk 18 above)
Comment (ZhennanQin) [reply python/mxnet/contrib/quantization.py:92]: Will fix.

File: src/executor/attach_op_execs_pass.cc (hunk 19 above)
Comment (ZhennanQin) [reply src/executor/attach_op_execs_pass.cc:162]: It's related to mkldnn stateful op. As mkldnn_conv is defined as a stateful op, we need this fix to avoid extra order.

File: src/ndarray/ndarray.cc (hunk 20 above)
Comment (ZhennanQin) [reply src/ndarray/ndarray.cc:452]: Moved this into another file which it should be.

File: src/operator/subgraph/mkldnn/mkldnn_conv.cc (hunk 22 above)
Comment (ZhennanQin) [reply src/operator/subgraph/mkldnn/mkldnn_conv.cc:163]: No, fwd will be changed.

File: src/operator/subgraph/mkldnn/mkldnn_conv.cc (hunk 23 above)
Comment (ZhennanQin) [reply src/operator/subgraph/mkldnn/mkldnn_conv.cc:422]: Yeah.

File: src/operator/subgraph/mkldnn/mkldnn_conv_property.cc (hunk 24 above)
Comment (ZhennanQin) [reply src/operator/subgraph/mkldnn/mkldnn_conv_property.cc]: For this version, yes.

File: src/operator/nn/mkldnn/mkldnn_convolution-inl.h [hunk 25]
```diff
@@ -35,19 +36,79 @@
...
+    DMLC_DECLARE_FIELD(with_bn).set_default(false)
+    .describe("Add post batchnorm.");
+    DMLC_DECLARE_FIELD(with_relu).set_default(false)
+    .describe("Add post relu");
+    DMLC_DECLARE_FIELD(with_sum).set_default(false)
+    .describe("Add post sum");
+    DMLC_DECLARE_FIELD(with_postsum_relu).set_default(false)
+    .describe("Add post relu after sum");
+    DMLC_DECLARE_FIELD(quantized).set_default(false)
+    .describe("enable quantization");
```
Comment (pengzhao-intel) [inline src/operator/nn/mkldnn/mkldnn_convolution-inl.h:62]: enable -> Enable

File: src/operator/nn/mkldnn/mkldnn_convolution-inl.h [hunk 26]
```diff
@@ -35,19 +36,79 @@
...
+  ConvolutionParam conv_param;
+  MKLDNNConvParam mkldnn_param;
+  float sum_scale;
+  std::vector<float> requantize_scales;
+};
+
+static inline bool IsOutputUInt8(const MKLDNNConvParam &mkldnn_param) {
+  return ((!mkldnn_param.with_sum) && mkldnn_param.with_relu) ||
+         mkldnn_param.with_postsum_relu;
+}
```
Comment (pengzhao-intel) [inline src/operator/nn/mkldnn/mkldnn_convolution-inl.h:97]: Add comments why select this condition

File: src/operator/nn/mkldnn/mkldnn_convolution.cc [hunk 27]
```diff
@@ -34,55 +34,98 @@
 namespace mxnet {
 namespace op {
 
+DMLC_REGISTER_PARAMETER(MKLDNNConvParam);
+
 bool SupportMKLDNNConv(const ConvolutionParam& params, const NDArray &input) {
   if (params.kernel.ndim() != 2)
     return false;
-  return input.dtype() == mshadow::kFloat32 && input.shape().ndim() == 4;
+  return SupportMKLDNNQuantize(input.dtype()) && input.shape().ndim() == 4;
```
Comment (pengzhao-intel) [inline src/operator/nn/mkldnn/mkldnn_convolution.cc:42]: avoid check Support Quantize in original FP32 path

File: src/operator/nn/mkldnn/mkldnn_convolution.cc [hunk 28]
```diff
@@ -34,55 +34,98 @@
...
+  strides[1] = param.conv_param.stride[1];
   mkldnn::memory::dims padding{0, 0};
-  padding[0] = param.pad[0];
-  padding[1] = param.pad[1];
-  if (param.dilate.ndim() == 0 && bias == nullptr) {
+  padding[0] = param.conv_param.pad[0];
+  padding[1] = param.conv_param.pad[1];
+  mkldnn::primitive_attr attr;
+  mkldnn::post_ops ops;
+  if (param.mkldnn_param.with_relu) {
```
Comment (pengzhao-intel) [inline src/operator/nn/mkldnn/mkldnn_convolution.cc:80]: are these options exclusive? or can be overlap? 
is `switch` or `if... else..` statement represent the logic better?

File: src/operator/nn/mkldnn/mkldnn_convolution.cc [hunk 29]
```diff
@@ -34,55 +34,98 @@
...
+  if (param.mkldnn_param.with_postsum_relu) {
+    float scale = 1.0f;            // for fp32, scale is 1.
+    float alpha = 0.0f;            // negative slope for mkldnn_eltwise_relu.
+    float beta = 1.0f;             // ignored for mkldnn_eltwise_relu.
+    ops.append_eltwise(scale, eltwise_relu, alpha, beta);
+  }
+  attr.set_post_ops(ops);
+
+  if (param.mkldnn_param.quantized) {
+    int mask = param.mkldnn_param.weight_channelwise_scale ? 2 : 0;
```
Comment (pengzhao-intel) [inline src/operator/nn/mkldnn/mkldnn_convolution.cc]: what's 2 and 0 mean? define an NAME?

File: src/operator/nn/mkldnn/mkldnn_convolution.cc [hunk 30]
```diff
@@ -238,17 +281,17 @@ MKLDNNConvForward &GetConvFwd(const nnvm::NodeAttrs& attrs, const bool is_train,
...
+                                         const std::vector<OpReqType> &req,
+                                         const std::vector<NDArray> &out_data) {
   TmpMemMgr::Get()->Init(ctx.requested[conv::kTempSpace]);
-  const ConvolutionParam& param = nnvm::get<ConvolutionParam>(attrs.parsed);
   NDArray weight = in_data[conv::kWeight];
-  MKLDNNConvForward &fwd = GetConvFwd(attrs, ctx.is_train, in_data[conv::kData], weight,
-      param.no_bias ? nullptr : &in_data[conv::kBias], out_data[conv::kOut]);
-
-  auto data_mem = in_data[conv::kData].GetMKLDNNDataReorder(fwd.fwd_pd.src_primitive_desc());
+  bool no_bias = param.conv_param.no_bias && !param.mkldnn_param.with_bn;
```
Comment (pengzhao-intel) [inline src/operator/nn/mkldnn/mkldnn_convolution.cc:292]: could we handle/wrapper fusion parameter in some mid-level?
so we don't need to check it explicitly.

File: src/operator/nn/mkldnn/mkldnn_convolution.cc [hunk 31]
```diff
@@ -257,32 +300,59 @@ void MKLDNNConvolutionForward(const nnvm::NodeAttrs& attrs, const OpContext &ctx
...
     } else {
       weight_mem = weight.GetMKLDNNData();
-      CHECK(weight_mem->get_primitive_desc() == fwd.fwd_pd.weights_primitive_desc());
+      CHECK(weight_mem->get_primitive_desc() == fwd->fwd_pd.weights_primitive_desc());
     }
   }
-  auto out_mem = CreateMKLDNNMem(out_data[conv::kOut], fwd.fwd_pd.dst_primitive_desc(),
-                                 req[conv::kOut]);
+  mkldnn_output_t out_mem;
+  if (param.mkldnn_param.with_sum) {
```
Comment (pengzhao-intel) [inline src/operator/nn/mkldnn/mkldnn_convolution.cc:320]: as previous comments

File: src/operator/quantization/mkldnn/mkldnn_quantize-inl.h [hunk 32]
```diff
@@ -75,6 +75,11 @@ static void MKLDNNQuantizeComputeKer(const std::vector<NDArray>& inputs,
   auto i_mpd = i_mem->get_primitive_desc();
   auto i_desc = i_mpd.desc();
   mkldnn::memory::format i_fmt = static_cast<mkldnn::memory::format>(i_desc.data.format);
+  if (i_fmt == mkldnn::memory::format::nchw ||
```
Comment (pengzhao-intel) [inline src/operator/quantization/mkldnn/mkldnn_quantize-inl.h:78]: add comments why

**Review by pengzhao-intel — COMMENTED** [review]
Look at conv
